### PR TITLE
DEVP-1139 sentry newrelic private channels

### PR DESCRIPTION
**Pull request by bh-anton-k**
```
$ ../shipcat/target/debug/shipcat values -r dev-uk SNIP
shipcat: values error: Private channel GSNIPSNIP is NOT supported as NewRelic target
```

### CODE REVIEW

File: shipcat_filebacked/src/newrelic_source.rs [hunk 1]
```diff
@@ -38,11 +38,15 @@ pub struct NewrelicSource {
...
     fn build(self, default_channel: &SlackChannel) -> Result<Option<Newrelic>> {
-        let slack = self
+        let slack: SlackChannel = self
             .slack
             .map(|s| s.verify().map(|_| s))
             .unwrap_or_else(|| Ok(default_channel.clone()))?;
 
+        if slack.starts_with("G") {
+            bail!("Private channels are NOT supported");
+        }
```
Comment (clux) [inline shipcat_filebacked/src/newrelic_source.rs]: as long as this validates in manifests locally, this is ok

File: shipcat_filebacked/src/sentry_source.rs [hunk 2]
```diff
@@ -31,6 +31,10 @@ impl Build<Sentry, SlackChannel> for SentrySource {
             .map(|s| s.verify().map(|_| s))
             .unwrap_or_else(|| Ok(default_channel.clone()))?;
 
+        if slack.starts_with("G") {
+            bail!("Private channels are NOT supported");
```
Comment (clux) [inline shipcat_filebacked/src/sentry_source.rs]: please include the name of the channel so it's not lost in the sea of validation errors

File: shipcat_filebacked/src/newrelic_source.rs (hunk 1 above)
Comment (clux) [reply shipcat_filebacked/src/newrelic_source.rs]: though please include the name of the channel so it's not lost in the sea of validation errors

File: shipcat_filebacked/src/newrelic_source.rs [hunk 3]
```diff
@@ -334,25 +350,52 @@ newrelic:
...
+    #[test]
+    fn test_private_slack_channel_from_default() -> Result<()> {
+        test_parse_and_merge_fails_with(TestSet {
+            manifest_yml: "newrelic: {}".into(),
+            env_yml: "{}".into(),
+            default_channel: "GDEFTEAM8".into(),
+            crd_expected: r###"---
+{}"###.into(),
+        },
+        "Private channels are NOT supported".as_ref())
```
Comment (clux) [inline shipcat_filebacked/src/newrelic_source.rs]: i wouldn't even bother to test this. changing the log message breaks the test? yuck.

File: shipcat_filebacked/src/newrelic_source.rs [hunk 4]
```diff
@@ -334,25 +350,52 @@ newrelic:
...
+  slack: G1DEVOPS8"###.into(),
+            env_yml: "{}".into(),
+            default_channel: "CDEFTEAM8".into(),
+            crd_expected: r###"---
+sentry:
+  slack: COVERRIDE
+  silent: true"###.into(),
+        },
+        "Private channels are NOT supported")
+    }
```
Comment (clux) [inline shipcat_filebacked/src/newrelic_source.rs]: are you testing merge logic here? why are you not satisfied with earlier merge logic tests?

File: shipcat_filebacked/src/sentry_source.rs [hunk 5]
```diff
@@ -68,6 +72,18 @@ mod tests {
...
 
+    impl TestSet {
+        fn build(&self) -> Result<Option<Sentry>> {
+            let manifest: ManifestOverridesNarrowed = serde_yaml::from_str(&self.manifest_yml)?;
+            let prod: ManifestOverridesNarrowed = serde_yaml::from_str(&self.env_yml)?;
+            let merge_with_env = manifest.merge(prod);
+
+            println!("merge result:\n{:#?}", merge_with_env);
+            let slack_default = &SlackChannel::new(&self.default_channel);
+            return merge_with_env.sentry.build(slack_default);
```
Comment (clux) [inline shipcat_filebacked/src/sentry_source.rs]: oh you are actually changing merge behavior, hm. why does this require a new builder? :thinking:

File: shipcat_filebacked/src/sentry_source.rs [hunk 6]
```diff
@@ -111,23 +127,50 @@ sentry:
...
+sentry:
+  slack: G1DEVOPS8"###.into(),
+            env_yml: "{}".into(),
+            default_channel: "CDEFTEAM8".into(),
+            crd_expected: r###"---
+sentry:
+  slack: COVERRIDE
+  silent: true"###.into(),
+        },
+        "Private channels are NOT supported")
```
Comment (clux) [inline shipcat_filebacked/src/sentry_source.rs]: bunch more of the same tests

File: shipcat_filebacked/src/newrelic_source.rs (hunk 1 above)
Comment (bh-anton-k) [reply shipcat_filebacked/src/newrelic_source.rs]: well here's the catch - we prefer to have messages formulated in a particular way buuuuut...

File: shipcat_filebacked/src/newrelic_source.rs (hunk 3 above)
Comment (bh-anton-k) [reply shipcat_filebacked/src/newrelic_source.rs]: ...buuuuut we prefer not to write the tests against those messages

this seems to be one of `who watches the watchers` examples - I'd nuke the tests from the orbit and switch to assume-the-best approach for basic stuff

File: shipcat_filebacked/src/sentry_source.rs (hunk 5 above)
Comment (bh-anton-k) [reply shipcat_filebacked/src/sentry_source.rs]: this is common test code I extracted both for happy and unhappy paths (it's not a `Build<...>` implementation) - as unhappy paths are not to be test-covered this is reverted as well

File: shipcat_definitions/src/lib.rs [hunk 7]
```diff
@@ -67,6 +67,10 @@ error_chain! {
             description("secret could not be reached or accessed")
             display("secret '{}'", &key)
         }
+        FailedToBuildManifest(service_name: String, region_name: String) {
+            description("failed to build manifest")
+            display("failed to build manifest for {} in {}", &service_name, &region_name)
+        }
```
Comment (clux) [inline shipcat_definitions/src/lib.rs]: Ah, nice, an actual error type :+1:

File: shipcat_cli/src/validate.rs [hunk 8]
```diff
@@ -21,8 +22,8 @@ pub fn regional_manifests(conf: &Config, reg: &Region) -> Result<()> {
         .partition_map(Either::from);
     if !errs.is_empty() {
         for e in &errs {
-            error!("{}", e);
-            debug!("{:?}", e);
+            error!("{}", e.display_chain());
+            debug!("{:?}", e.display_chain());
```
Comment (clux) [inline shipcat_cli/src/validate.rs]: display_chain is a type with a similar Display and Debug impl?

Comment (bh-anton-k) [reply shipcat_cli/src/validate.rs]:
```shipcat::validate: Error: failed to build manifest for SNAP in sn-ap
Caused by: Private/personal channel @USNAP is NOT supported as Sentry target

shipcat::validate: DisplayChain(Error(Mani(Error(FailedToBuildManifest("SNAP", "sn-ap"), State { next_error: Some(Error(Msg("Private/personal channel @USNAP is NOT supported as Sentry target"), State { next_error: None, backtrace: InternalBacktrace { backtrace: None } })), backtrace: InternalBacktrace { backtrace: None } })), State { next_error: None, backtrace: InternalBacktrace { backtrace: None } }))```

Comment (clux) [reply shipcat_cli/src/validate.rs]: Sensible!

**Pull request merged**